ip_proofs/src/tipa/mod.rs

```diff
@@ -0,0 +1,499 @@
+        com: (&LMC::Output, &RMC::Output, &IPC::Output),
+        proof: &TIPAProof<IP, LMC, RMC, IPC, P, D>,
+    ) -> Result<bool, Error> {
+        Self::verify_with_srs_shift(v_srs, ck_t, com, proof, &<P::Fr>::one())
+    }
+
+    pub fn verify_with_srs_shift(
+        v_srs: &VerifierSRS<P>,
+        ck_t: &IPC::Key,
+        com: (&LMC::Output, &RMC::Output, &IPC::Output),
+        proof: &TIPAProof<IP, LMC, RMC, IPC, P, D>,
+        r_shift: &P::Fr,
+    ) -> Result<bool, Error> {
+        let (base_com, transcript) =
+            GIPA::verify_recursive_challenge_transcript(com, &proof.gipa_proof)?;
+        let transcript_inverse = transcript.iter().map(|x| x.inverse().unwrap()).collect();
+
+        // Verify commitment keys wellformed
+        let (ck_a_final, ck_b_final) = &proof.final_ck;
+        let (ck_a_proof, ck_b_proof) = &proof.final_ck_proof;
+
+        // KZG challenge point
+        let mut counter_nonce: usize = 0;
+        let c = loop {
+            let mut hash_input = Vec::new();
+            hash_input.extend_from_slice(&counter_nonce.to_be_bytes()[..]);
+            //TODO: Should use CanonicalSerialize instead of ToBytes
+            hash_input.extend_from_slice(&to_bytes![
+                transcript.first().unwrap(),
+                ck_a_final,
+                ck_b_final
+            ]?);
+            if let Some(c) = LMC::Scalar::from_random_bytes(&D::digest(&hash_input)) {
+                break c;
+            };
+            counter_nonce += 1;
+        };
+
+        let ck_a_polynomial_c_eval = polynomial_evaluation_product_form_from_transcript(
+            &transcript_inverse,
+            &c,
+            &r_shift.inverse().unwrap(),
+        );
+        let ck_b_polynomial_c_eval =
+            polynomial_evaluation_product_form_from_transcript(&transcript, &c, &<P::Fr>::one());
+
+        let ck_a_valid =
+            P::pairing(
+                v_srs.g.clone(),
+                ck_a_final.clone() - &v_srs.h.mul(ck_a_polynomial_c_eval),
+            ) == P::pairing(v_srs.g_beta.clone() - &v_srs.g.mul(c), ck_a_proof.clone());
+        let ck_b_valid =
+            P::pairing(
+                ck_b_final.clone() - &v_srs.g.mul(ck_b_polynomial_c_eval),
+                v_srs.h.clone(),
+            ) == P::pairing(ck_b_proof.clone(), v_srs.h_alpha.clone() - &v_srs.h.mul(c));
+
+        // Verify base inner product commitment
+        let (com_a, com_b, com_t) = base_com;
+        let a_base = vec![proof.gipa_proof.r_base.0.clone()];
+        let b_base = vec![proof.gipa_proof.r_base.1.clone()];
+        let t_base = vec![IP::inner_product(&a_base, &b_base)?];
+        let base_valid = LMC::verify(&vec![ck_a_final.clone()], &a_base, &com_a)?
+            && RMC::verify(&vec![ck_b_final.clone()], &b_base, &com_b)?
+            && IPC::verify(&vec![ck_t.clone()], &t_base, &com_t)?;
+
+        Ok(ck_a_valid && ck_b_valid && base_valid)
+    }
+}
+
+pub fn structured_generators_scalar_power<G: Group>(
```

Comment:
If not using the existing KZG10 impl, then you should use https://github.com/scipr-lab/zexe/blob/38b6f6f2c6b9a6a1b5e8eb90b1287c98b06520d1/algebra-core/src/msm/fixed_base.rs

Reply:
:+1:  Changed parameterization to `ProjectiveGroup`